src/charm.py

```diff
@@ -43,6 +46,74 @@ def __init__(self, *args: typing.Any) -> None:
         super().__init__(*args)
         self.framework.observe(self.on.config_changed, self.reconciliation)
         self.framework.observe(self.on.spring_boot_app_pebble_ready, self.reconciliation)
+        self.database: typing.Optional[DatabaseRequires] = self._setup_database_requirer(
+            "mysql", "mysql"
+        )
+
+    def _setup_database_requirer(self, relation_name: str, database_name: str) -> DatabaseRequires:
+        """Handle the creation of relations and listeners.
+
+        Args:
+            relation_name: Name of the data relation
+            database_name: Name of the database (could be overwritten by the provider)
+
+        Returns:
+            DatabaseRequires object
+        """
+        database_requirer = DatabaseRequires(
+            self,
+            relation_name=relation_name,
+            database_name=database_name,
+        )
+        self.framework.observe(database_requirer.on.database_created, self._on_database_created)
+        self.framework.observe(self.on[relation_name].relation_broken, self._on_relation_broken)
+        return database_requirer
+
+    def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application.
+
+        Args:
+            event: The DatabaseCreatedEvent object
+        """
+        logger.debug("Database credentials are received: %s", event.username)
+        self.reconciliation(event)
+
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle relation broken event.
+
+        Args:
+            event: The RelationBrokenEvent object
+        """
+        logger.debug("Relation with database removed: %s", event.relation.name)
+        self.reconciliation(event)
```

Comment:
Not 100% convinced about it, but we should stay consistent.
Either we define a method for every event, even if they only log and call reconciliation, or we directly bind reconciliation in the `observe`

Here we're doing both as Weii did the second way and it's done the first way here.
@weiiwang01 what's your opinion ?

Reply:
Actually, this way of doing things is only useful if we want the logger output.  
I'm okay with directly binding reconciliation()

Reply:
Well we probably could still log it in the reconciliation method, don't know if you can get the real type in python from its instance, or if the event contains some field that specifies its type

Reply:
The `reconciliation` method already logs the event that trigger the reconciliation. So the extra log is not necessary and direct binding is okay.

Reply:
Then I'm more in favor of a direct binding :)

Reply:
Changed